pkg/provider/client.go

```diff
@@ -20,64 +20,66 @@ import (
+	kubernetesenv "github.com/nutanix-cloud-native/prism-go-client/environment/providers/kubernetes"
+	envtypes "github.com/nutanix-cloud-native/prism-go-client/environment/types"
+	prismclientv3 "github.com/nutanix-cloud-native/prism-go-client/v3"
 	"k8s.io/client-go/informers"
 	coreinformers "k8s.io/client-go/informers/core/v1"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2"
 
+	"github.com/nutanix-cloud-native/cloud-provider-nutanix/internal/constants"
 	"github.com/nutanix-cloud-native/cloud-provider-nutanix/pkg/provider/config"
 	"github.com/nutanix-cloud-native/cloud-provider-nutanix/pkg/provider/interfaces"
 )
 
 const errEnvironmentNotReady = "environment not initialized or ready yet"
 
 type nutanixClient struct {
-	env               *envTypes.Environment
+	env               envtypes.Environment
 	config            config.Config
 	secretInformer    coreinformers.SecretInformer
 	sharedInformers   informers.SharedInformerFactory
 	configMapInformer coreinformers.ConfigMapInformer
+	clientCache       *prismclientv3.ClientCache
 }
 
-func (n *nutanixClient) Get() (interfaces.Prism, error) {
-	if err := n.setupEnvironment(); err != nil {
-		return nil, fmt.Errorf("%s: %v", errEnvironmentNotReady, err)
+// Key returns the constant client name
+// This implements the CachedClientParams interface of prism-go-client
+func (n *nutanixClient) Key() string {
+	return constants.ClientName
+}
+
+// ManagementEndpoint returns the management endpoint of the Nutanix cluster
+// This implements the CachedClientParams interface of prism-go-client
+func (n *nutanixClient) ManagementEndpoint() envtypes.ManagementEndpoint {
+	if n.env == nil {
+		klog.Error("environment not initialized")
+		return envtypes.ManagementEndpoint{}
 	}
-	env := *n.env
-	me, err := env.GetManagementEndpoint(envTypes.Topology{})
+
+	mgmtEndpoint, err := n.env.GetManagementEndpoint(envtypes.Topology{})
 	if err != nil {
-		return nil, err
-	}
-	creds := &prismgoclient.Credentials{
-		URL:      me.Address.Host, // Not really an URL
-		Endpoint: me.Address.Host,
-		Insecure: me.Insecure,
-		Username: me.ApiCredentials.Username,
-		Password: me.ApiCredentials.Password,
+		klog.Errorf("failed to get management endpoint: %s", err.Error())
+		return envtypes.ManagementEndpoint{}
 	}
 
-	clientOpts := make([]prismClientV3.ClientOption, 0)
-	if me.AdditionalTrustBundle != "" {
-		clientOpts = append(clientOpts, prismClientV3.WithPEMEncodedCertBundle([]byte(me.AdditionalTrustBundle)))
-	}
+	return *mgmtEndpoint
+}
 
-	nutanixClient, err := prismClientV3.NewV3Client(*creds, clientOpts...)
-	if err != nil {
-		return nil, err
+func (n *nutanixClient) Get() (interfaces.Prism, error) {
+	if err := n.setupEnvironment(); err != nil {
+		return nil, fmt.Errorf("%s: %w", errEnvironmentNotReady, err)
 	}
 
-	_, err = nutanixClient.V3.GetCurrentLoggedInUser(context.Background())
+	client, err := n.clientCache.GetOrCreate(n)
```

Comment:
I think we wont be deleting it from the cache. also there will always be only one instance in the cache hopefully given single cluster

Reply:
correct. this is to just offload client creation to an existing method in the client.